From b095e73cd6c9dafa15ec06ddc41a2b717f19b98e Mon Sep 17 00:00:00 2001
From: Raphael Mounier <mounierr07@gmail.com>
Date: Thu, 6 Oct 2022 08:37:44 +0200
Subject: [PATCH 6/7] Fix android.hardware.radio@1.1/1.0, which don't provide a
 slot-port mapping index

Google broke radio 1.1/1.0 support in I82587aa44db4cb2ad7cd708b51ead00d678be12e
bring its support back
---
 .../android/internal/telephony/uicc/UiccCard.java  |  4 +++-
 .../internal/telephony/uicc/UiccController.java    |  9 +++++++--
 .../android/internal/telephony/uicc/UiccPort.java  |  9 +++++++--
 .../android/internal/telephony/uicc/UiccSlot.java  | 14 +++++++++-----
 4 files changed, 26 insertions(+), 10 deletions(-)

diff --git a/src/java/com/android/internal/telephony/uicc/UiccCard.java b/src/java/com/android/internal/telephony/uicc/UiccCard.java
index 689e4b7be9..a316a8142f 100644
--- a/src/java/com/android/internal/telephony/uicc/UiccCard.java
+++ b/src/java/com/android/internal/telephony/uicc/UiccCard.java
@@ -104,7 +104,9 @@ public class UiccCard {
             mCardState = ics.mCardState;
             updateCardId(ics.iccid);
             if (mCardState != CardState.CARDSTATE_ABSENT) {
-                int portIdx = ics.mSlotPortMapping.mPortIndex;
+                int portIdx = 0;
+                if (ics.mSlotPortMapping != null)
+                    portIdx = ics.mSlotPortMapping.mPortIndex;
                 UiccPort port = mUiccPorts.get(portIdx);
                 if (port == null) {
                     if (this instanceof EuiccCard) {
diff --git a/src/java/com/android/internal/telephony/uicc/UiccController.java b/src/java/com/android/internal/telephony/uicc/UiccController.java
index 70cfce00b6..d8fe8fa34e 100644
--- a/src/java/com/android/internal/telephony/uicc/UiccController.java
+++ b/src/java/com/android/internal/telephony/uicc/UiccController.java
@@ -762,7 +762,9 @@ public class UiccController extends Handler {
 
         logWithLocalLog("onGetIccCardStatusDone: phoneId " + index + " IccCardStatus: " + status);
 
-        int slotId = status.mSlotPortMapping.mPhysicalSlotIndex;
+        int slotId = INVALID_SLOT_ID;
+        if (status.mSlotPortMapping != null)
+            slotId = status.mSlotPortMapping.mPhysicalSlotIndex;
         if (VDBG) log("onGetIccCardStatusDone: phoneId " + index + " physicalSlotIndex " + slotId);
         if (slotId == INVALID_SLOT_ID) {
             slotId = index;
@@ -799,7 +801,10 @@ public class UiccController extends Handler {
         if (isEuicc) {
             cardString = ((EuiccCard) card).getEid();
         } else {
-            cardString = card.getUiccPort(status.mSlotPortMapping.mPortIndex).getIccId();
+            if (status.mSlotPortMapping == null)
+                cardString = card.getUiccPort(0).getIccId();
+            else
+                cardString = card.getUiccPort(status.mSlotPortMapping.mPortIndex).getIccId();
         }
 
         if (cardString != null) {
diff --git a/src/java/com/android/internal/telephony/uicc/UiccPort.java b/src/java/com/android/internal/telephony/uicc/UiccPort.java
index 0152dda088..93f3401d3f 100644
--- a/src/java/com/android/internal/telephony/uicc/UiccPort.java
+++ b/src/java/com/android/internal/telephony/uicc/UiccPort.java
@@ -74,8 +74,13 @@ public class UiccPort {
             mContext = c;
             mCi = ci;
             mIccid = ics.iccid;
-            mPortIdx = ics.mSlotPortMapping.mPortIndex;
-            mPhysicalSlotIndex = ics.mSlotPortMapping.mPhysicalSlotIndex;
+            if (ics.mSlotPortMapping != null) {
+                mPortIdx = ics.mSlotPortMapping.mPortIndex;
+                mPhysicalSlotIndex = ics.mSlotPortMapping.mPhysicalSlotIndex;
+	    } else {
+	        mPortIdx = 0;
+                mPhysicalSlotIndex = 0;
+	    }
             if (mUiccProfile == null) {
                 mUiccProfile = TelephonyComponentFactory.getInstance()
                         .inject(UiccProfile.class.getName()).makeUiccProfile(
diff --git a/src/java/com/android/internal/telephony/uicc/UiccSlot.java b/src/java/com/android/internal/telephony/uicc/UiccSlot.java
index 9b5b315b56..cc3598a92e 100644
--- a/src/java/com/android/internal/telephony/uicc/UiccSlot.java
+++ b/src/java/com/android/internal/telephony/uicc/UiccSlot.java
@@ -110,10 +110,14 @@ public class UiccSlot extends Handler {
     public void update(CommandsInterface ci, IccCardStatus ics, int phoneId, int slotIndex) {
         if (DBG) log("cardStatus update: " + ics.toString());
         synchronized (mLock) {
-            mPortIdxToPhoneId.put(ics.mSlotPortMapping.mPortIndex, phoneId);
+            int portIndex = 0;
+            if (ics.mSlotPortMapping != null) {
+                portIndex = ics.mSlotPortMapping.mPortIndex;
+            }
+            mPortIdxToPhoneId.put(portIndex, phoneId);
             CardState oldState = mCardState;
             mCardState = ics.mCardState;
-            mIccIds.put(ics.mSlotPortMapping.mPortIndex, ics.iccid);
+            mIccIds.put(portIndex, ics.iccid);
             parseAtr(ics.atr);
             mIsRemovable = isSlotRemovable(slotIndex);
 
@@ -124,7 +128,7 @@ public class UiccSlot extends Handler {
 
             if (absentStateUpdateNeeded(oldState)) {
                 updateCardStateAbsent(ci.getRadioState(), phoneId,
-                        ics.mSlotPortMapping.mPortIndex);
+                        portIndex);
             // Because mUiccCard may be updated in both IccCardStatus and IccSlotStatus, we need to
             // create a new UiccCard instance in two scenarios:
             //   1. mCardState is changing from ABSENT to non ABSENT.
@@ -133,7 +137,7 @@ public class UiccSlot extends Handler {
                     || mUiccCard == null) && mCardState != CardState.CARDSTATE_ABSENT) {
                 // No notification while we are just powering up
                 if (radioState != TelephonyManager.RADIO_POWER_UNAVAILABLE
-                        && mLastRadioState.getOrDefault(ics.mSlotPortMapping.mPortIndex,
+                        && mLastRadioState.getOrDefault(portIndex,
                         TelephonyManager.RADIO_POWER_UNAVAILABLE)
                         != TelephonyManager.RADIO_POWER_UNAVAILABLE) {
                     if (DBG) log("update: notify card added");
@@ -164,7 +168,7 @@ public class UiccSlot extends Handler {
                     mUiccCard.update(mContext, ci, ics, phoneId);
                 }
             }
-            mLastRadioState.put(ics.mSlotPortMapping.mPortIndex, radioState);
+            mLastRadioState.put(portIndex, radioState);
         }
     }
 
-- 
2.25.1

